js/src/scene.ts

```diff
@@ -0,0 +1,436 @@
+
+export class SceneModel extends DOMWidgetModel {
+    defaults(): any {
+        return {
+            ...super.defaults(),
+            _model_name: 'SceneModel',
+            _view_name: 'SceneView',
+            _model_module_version: semver,
+            _view_module_version: semver,
+            _model_module: 'exatomic',
+            _view_module: 'exatomic',
+            width: 200,
+            height: 200,
+        }
+    }
+}
+
+export class SceneView extends DOMWidgetView {
+    scene: three.Scene
+
+    camera: three.PerspectiveCamera
+
+    renderer: three.WebGLRenderer | null
+
+    controls: any // TrackBallControls
+
+    raycaster: three.Raycaster
+
+    mouse: three.Vector2
+
+    hudscene: three.Scene
+
+    hudcamera: three.OrthographicCamera
+
+    hudcanvas: HTMLCanvasElement
+
+    hudcontext: CanvasRenderingContext2D | null
+
+    hudtexture: three.Texture
+
+    hudsprite: three.Sprite
+
+    hudfontsize: number
+
+    promises: Promise<any>
+
+    selected: three.Mesh[]
+
+    initialize(parameters: any): void {
+        super.initialize(parameters)
+        this.initListeners()
+        this.selected = []
+        this.hudfontsize = 24
+        this.promises = this.init()
+        this.displayed.then(() => {
+            this.resize()
+            this.setCameraFromScene()
+        })
+    }
+
+    init(): Promise<any> {
+        /* """
+        init
+        ---------------
+        Promise it will work
+        */
+        window.addEventListener('resize', this.resize.bind(this))
+        return Promise.all([
+            this.initScene(),
+            this.initRenderer(),
+            this.initHud(),
+        ]).then(() => {
+            this.initObj()
+            this.finalizeHudcanvas()
+            this.finalizeInteractive()
+        })
+    }
+
+    initScene() {
+        /* """
+        initScene
+        ---------------
+        An opinionated three.js Scene 
+        */
+        const scene = new three.Scene()
+        const amlight = new three.AmbientLight(0xdddddd, 0.5)
+        const dlight0 = new three.DirectionalLight(0xdddddd, 0.3)
+        const sunlight = new three.SpotLight(0xdddddd, 0.3, 0, Math.PI / 2)
+        const shadowcam = new three.PerspectiveCamera(30, 1, 1500, 5000)
+        dlight0.position.set(-1000, -1000, -1000)
+        sunlight.position.set(1000, 1000, 1000)
+        sunlight.castShadow = true
+        sunlight.shadow = new three.SpotLightShadow(shadowcam)
+        //    camera: shadowcam,
+        //    isSpotLightShadow: true,
+        //})
+        sunlight.shadow.bias = 0.0
+        scene.add(amlight)
+        scene.add(dlight0)
+        scene.add(sunlight)
+        return Promise.resolve(scene).then((scn) => {
+            this.scene = scn
+        })
+    }
+
+    initRenderer(): any {
+        /* """
+        initRenderer
+        ---------------
+        A WebGLRenderer, PerspectiveCamera and
+        TrackBallControls object.
+        */
+        return Promise.all([
```

Comment:
This is a pretty long chain and its a bit hard to tell how to debug if the promises aren't fulfilled. It might be worth adding the optional arg to trace/log/inform when something fails here.

Reply:
This should be addressed with recent changes. We should probably still handle rejection of the single promise that is left.